src/Http/Http.Extensions/test/ParameterBindingMethodCacheTests.cs

```diff
@@ -274,6 +274,63 @@ public void FindBindAsyncMethod_FindsNonNullableReturningBindAsyncMethodGivenNul
             Assert.True(new ParameterBindingMethodCache().HasBindAsyncMethod(parameterInfo));
         }
 
+        [Theory]
+        [InlineData(typeof(InvalidVoidReturnTryParseStruct))]
+        [InlineData(typeof(InvalidVoidReturnTryParseClass))]
+        [InlineData(typeof(InvalidWrongTypeTryParseStruct))]
+        [InlineData(typeof(InvalidWrongTypeTryParseClass))]
+        [InlineData(typeof(InvalidTryParseNullableStruct))]
+        [InlineData(typeof(InvalidTooFewArgsTryParseStruct))]
+        [InlineData(typeof(InvalidTooFewArgsTryParseClass))]
+        [InlineData(typeof(InvalidNonStaticTryParseStruct))]
+        [InlineData(typeof(InvalidNonStaticTryParseClass))]
+        public void FindTryParseMethod_ThrowsIfInvalidTryParseOnType(Type type)
+        {
+            var ex = Assert.Throws<InvalidOperationException>(
+                () => new ParameterBindingMethodCache().FindTryParseMethod(type));
+            Assert.StartsWith($"TryParse method found on {type.Name} with incorrect format. Must be a static method with format", ex.Message);
+            Assert.Contains($"bool TryParse(string, IFormatProvider, out {type.Name})", ex.Message);
+            Assert.Contains($"bool TryParse(string, out {type.Name})", ex.Message);
+        }
+
+        [Theory]
+        [InlineData(typeof(TryParseClassWithGoodAndBad))]
+        [InlineData(typeof(TryParseStructWithGoodAndBad))]
+        public void FindTryParseMethod_IgnoresInvalidTryParseIfGoodOneFound(Type type)
+        {
+            var method = new ParameterBindingMethodCache().FindTryParseMethod(type);
+            Assert.NotNull(method);
+        }
+
+        [Theory]
+        [InlineData(typeof(InvalidWrongReturnBindAsyncStruct))]
+        [InlineData(typeof(InvalidWrongReturnBindAsyncClass))]
+        [InlineData(typeof(InvalidWrongParamBindAsyncStruct))]
+        [InlineData(typeof(InvalidWrongParamBindAsyncClass))]
+        public void FindBindAsyncMethod_ThrowsIfInvalidBindAsyncOnType(Type type)
+        {
+            var cache = new ParameterBindingMethodCache();
+            var parameter = new MockParameterInfo(type, "anything");
+            var ex = Assert.Throws<InvalidOperationException>(
+                () => cache.FindBindAsyncMethod(parameter));
+            Assert.StartsWith($"BindAsync method found on {type.Name} with incorrect format. Must be a static method with format", ex.Message);
```

Comment:
How does this look with generic types? Do we want to use TypeNameHelper?

Reply:
Changed to use TypeNameHelper